File: app/src/main/java/com/veganbeauty/app/features/home/adapter/HomeTopSearchAdapter.java
package com.veganbeauty.app.features.home.adapter;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.veganbeauty.app.data.local.entities.ProductEntity;
import com.veganbeauty.app.databinding.HomeItemTopSearchBinding;

public class HomeTopSearchAdapter extends ListAdapter<ProductEntity, HomeTopSearchAdapter.ViewHolder> {

    public interface OnItemClickListener {
        void onItemClick(ProductEntity product);
    }

    private final OnItemClickListener onItemClick;

    public HomeTopSearchAdapter() {
        this(product -> {});
    }

    public HomeTopSearchAdapter(OnItemClickListener onItemClick) {
        super(new DiffCallback());
        this.onItemClick = onItemClick;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        HomeItemTopSearchBinding binding = HomeItemTopSearchBinding.inflate(
                LayoutInflater.from(parent.getContext()), parent, false
        );
        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.bind(getItem(position), position + 1);
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        private final HomeItemTopSearchBinding binding;

        public ViewHolder(HomeItemTopSearchBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        public void bind(ProductEntity product, int rank) {
            binding.getRoot().setOnClickListener(v -> onItemClick.onItemClick(product));
            binding.tvRank.setText(String.valueOf(rank));
            binding.tvProductName.setText(product.getName());
            
            com.veganbeauty.app.utils.ProductImageHelper.loadProductImage(binding.ivProduct, product);
        }
    }

    private static class DiffCallback extends DiffUtil.ItemCallback<ProductEntity> {
        @Override
        public boolean areItemsTheSame(@NonNull ProductEntity oldItem, @NonNull ProductEntity newItem) {
            return oldItem.getId().equals(newItem.getId());
        }

        @Override
        public boolean areContentsTheSame(@NonNull ProductEntity oldItem, @NonNull ProductEntity newItem) {
            return oldItem.equals(newItem);
        }
    }
}

File: app/src/main/java/com/veganbeauty/app/features/home/adapter/HomeTopSearchAdapter.kt
package com.veganbeauty.app.features.home.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import coil.load
import com.veganbeauty.app.data.local.entities.ProductEntity
import com.veganbeauty.app.databinding.HomeItemTopSearchBinding

class HomeTopSearchAdapter(
    private val onItemClick: (ProductEntity) -> Unit = {}
) : ListAdapter<ProductEntity, HomeTopSearchAdapter.ViewHolder>(DiffCallback()) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding = HomeItemTopSearchBinding.inflate(
            LayoutInflater.from(parent.context), parent, false
        )
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(getItem(position), position + 1)
    }

    inner class ViewHolder(
        private val binding: HomeItemTopSearchBinding
    ) : RecyclerView.ViewHolder(binding.root) {

        fun bind(product: ProductEntity, rank: Int) {
            binding.root.setOnClickListener { onItemClick(product) }
            binding.tvRank.text = rank.toString()
            binding.tvProductName.text = product.name
            binding.ivProduct.load(product.mainImage) {
                crossfade(true)
                placeholder(android.R.color.darker_gray)
            }
        }
    }

    private class DiffCallback : DiffUtil.ItemCallback<ProductEntity>() {
        override fun areItemsTheSame(oldItem: ProductEntity, newItem: ProductEntity) =
            oldItem.id == newItem.id

        override fun areContentsTheSame(oldItem: ProductEntity, newItem: ProductEntity) =
            oldItem == newItem
    }
}
